components/style/matching.rs

```diff
@@ -318,23 +347,18 @@ trait PrivateMatchMethods: TElement {
 
         // Grab the inherited values.
         let parent_el;
-        let parent_data;
+        let element_and_values; // So parent_el and style_to_inherit_from are known live.
         let style_to_inherit_from = match cascade_target {
             CascadeTarget::Normal => {
-                parent_el = self.inheritance_parent();
-                parent_data = parent_el.as_ref().and_then(|e| e.borrow_data());
-                let parent_style = parent_data.as_ref().map(|d| {
-                    // Sometimes Gecko eagerly styles things without processing
-                    // pending restyles first. In general we'd like to avoid this,
-                    // but there can be good reasons (for example, needing to
-                    // construct a frame for some small piece of newly-added
-                    // content in order to do something specific with that frame,
-                    // but not wanting to flush all of layout).
-                    debug_assert!(cfg!(feature = "gecko") ||
-                                  parent_el.unwrap().has_current_styles(d));
-                    d.styles.primary()
-                });
-                parent_style.map(|s| cascade_visited.values(s))
+                let info = match parent_info {
+                    Some(element_and_values) => element_and_values,
+                    None => {
+                        element_and_values = self.get_inherited_values_and_parent();
```

Comment:
We do end up going down this path, yes.  For example, get_after_change_style calls us with `CasacadeTarget::Normal` and a None for the `parent_info`.